cms/djangoapps/contentstore/courseware_index.py

```diff
@@ -0,0 +1,192 @@
+                this will apply downwards until overridden while recursively processing
+                the children
+
+            skip_index - simply walk the children in the tree, the content change is
+                older than the REINDEX_AGE window and would have been already indexed.
+                This should really only be passed from the recursive child calls when
+                this method has determined that it is safe to do so
+            """
+            is_indexable = hasattr(item, "index_dictionary")
+            item_index_dictionary = item.index_dictionary() if is_indexable else None
+            # if it's not indexable and it does not have children, then ignore
+            if not item_index_dictionary and not item.has_children:
+                return
+
+            item_id = unicode(item.scope_ids.usage_id)
+            indexed_items.add(item_id)
+
+            # if it has a defined start, then apply it and to it's children
+            if item.start and (not current_start_date or item.start > current_start_date):
+                current_start_date = item.start
+
+            if item.has_children:
+                # determine if it's okay to skip adding the children herein based upon how recently any may have changed
+                skip_child_index = skip_index or \
+                    (triggered_at is not None and (triggered_at - item.subtree_edited_on) > reindex_age)
+                for child_item in item.get_children():
+                    index_item(child_item, current_start_date, skip_index=skip_child_index)
+
+            if skip_index or not item_index_dictionary:
+                return
+
+            item_index = {}
+            # if it has something to add to the index, then add it
+            try:
+                item_index.update(location_info)
+                item_index.update(item_index_dictionary)
+                item_index['id'] = item_id
+                if current_start_date:
+                    item_index['start_date'] = current_start_date
+
+                searcher.index(DOCUMENT_TYPE, item_index)
+                indexed_count["count"] += 1
+            except Exception as err:  # pylint: disable=broad-except
+                # broad exception so that index operation does not fail on one item of many
+                log.warning('Could not index item: %s - %r', item.location, err)
+                error_list.append(_('Could not index item: {}').format(item.location))
+
+        def remove_deleted_items():
+            """
+            remove any item that is present in the search index that is not present in updated list of indexed items
+            as we find items we can shorten the set of items to keep
+            """
+            response = searcher.search(
+                doc_type=DOCUMENT_TYPE,
+                field_dictionary={"course": unicode(course_key)},
+                exclude_ids=indexed_items
```

Comment:
@martynjames I'm a bit concerned of how it would behave for actual course with many XBlocks. If I understand right, under the hood it would build an enormous `not - filter - or` query with number of clauses equal to number of used Xblocks, which might well be in hundreds. I believe elastic will swallow it just fine, but it would be nice to have a better proof than belief (load test ideally, but a couple of measurements with big courses would still suffice).

Also, the deletion approach is not the most efficient one as well - it deletes one document at a time, while there are more efficient approach built in: [delete by query](http://www.elastic.co/guide/en/elasticsearch/reference/0.90/docs-delete-by-query.html) (we're using 0.90 in devstack and prodstack, right?). It would be nice to have an ability to use bulk operations with search (where it makes sense). Not insisting on it though, and not a blocker for this PR.